### ISSUE TITLE
What is the environment needed for running this?

### ISSUE THREAD

**Issue by sebastiani** [xu-ji/IIC#4]
I'm having quite a bit of trouble running this, from simple stuff between python2.7 and 3.6 to weird behavior from my version of PyTorch not broadcasting types properly.

**Comment by xu-ji** [comment]
Hi, I've listed all the package versions here: https://github.com/xu-ji/IIC/blob/master/package_versions.txt

I used virtualenv to isolate the environment, highly recommend it.

**Comment by sebastiani** [comment]
I'm working on porting this to python3.6 and torch 1.0.1, I can make a PR if you think it will be useful.

**Comment by xu-ji** [comment]
I would like the results to be exactly reproducible. I think the simplest thing would be to keep the code separate. In the main readme for this repo, I'll add a list of links to associated repos for people check out.

(Btw I think you added some *pyc files to your repo, not sure if this was intentional)

**Comment by sebastiani** [comment]
Alright, I'll keep that in mind, right now I'm running a few of the examples using the translated code. Just out of curiosity, how long did it take to complete training for STL10(569)?

(Also, thanks for letting me know about the *.pyc's, I didn't notice!)

**Comment by xu-ji** [comment]
The most dramatic learning curve is in the first 30~50 epochs. It took 28 epochs for STL10 569 to go from 11% acc (random init) to 50% acc. This took 22 hours (including ~5min eval after every epoch). In total best performance was reached in just under 9 days.

**Comment by sebastiani** [comment]
Hey, thanks for the numbers! I left it running and accidentally forgot to check it at epoch 50, however at epoch 138 I got an average accuracy of 53.46% with the best at 56%. I'm assuming those numbers are reasonable judging from what you said. I'll test segmentation now(COCO-Stuff-3 (555)).

**Comment by Jian-danai** [comment]
How to install javapackages==1.0.0?

**Comment by DonkeyShot21** [comment]
Hi, sorry for reopening this. I am trying to reproduce your results using your code. In particular, I want to train model 640 (CIFAR 10, sobel, two heads) using a Tesla K40. It takes approx 1h20min for each epoch which is a lot. I tried using multiple GPUs (K40) but it didn't help, it looked even slower. Also, I downloaded your trained models and I saw you need more or less 1k epochs in order for the model to properly converge. Doing the maths it looks like it will be done in some months. What hardware did you use? Did you train on multiple GPUs?

**Comment by xu-ji** [comment]
I just uploaded some [logs](https://github.com/xu-ji/IIC/blob/master/examples/logs.tar.gz). This model was trained on 1 Tesla M40 12GB (~20 min per epoch) and then 1 Tesla P40 24GB (~15 min per epoch). "Per epoch" here includes both heads and evaluation.

From the config the best subhead reached max 60.16% accuracy by 300th epoch, 61.54% by 600th epoch, 61.72% at 861st epoch. (NB these values in the logs are using accuracy to pick the best sub-head.) So using 20 minutes per epoch, best performance takes (861 * 20)/(24*60) or 12 days, and reaching ~60% accuracy takes 4 days.

Using more GPUs and getting noticeably slower performance really should not be happening, it suggests a networking issue. But it is also not reasonable to expect epoch speedup unless you correspondingly increase the batch size used by the script, not sure if you are doing that.

**Comment by DonkeyShot21** [comment]
Thanks for the quick reply. I tried with 2x1080ti, I get under 10min per epoch, with a single 1080ti between 12 and 13min. For some reason, K40s are not good, or maybe the machine I was using has some issues. 

I did not increase the batch size because I thought it could hurt performance. How did you choose to use 660 for CIFAR10? Hardware constraints or tuning?

**Comment by xu-ji** [comment]
Mostly hardware constraints, as in general bigger batch size is better (up to some point). But it's not a perfectly monotonic relationship because of the learning rate so I did try a few values.

The larger the batch the better it approximates the whole dataset and the better it feeds the large number of clusters in the overclustering head. As the 660 number includes 3x augmentation (`num_dataloaders` variable) this is really 220 original images (each one duplicated 3 times and paired with a different random transform) per batch. From what I remember, both made a difference but the number of original images mattered more than the number of augmentations/nominal batch size.
